### PR TITLE
PGN parsing fixes

### PR DESCRIPTION
**Pull request by mbutsykin**
- fixes parsing of pgns with unescaped backslashes in linebreaks
- tolerance for positions with initial move number `0`. Looks like old chessbase (<12) in some cases can put `0.` as a move number after board setup...

### CODE REVIEW

**Review by yo35 — CHANGES_REQUESTED** [review]

File: src/private_position/fen.js [hunk 1]
```diff
@@ -288,7 +288,7 @@ exports.parseFEN = function(variant, fen, strict) {
 	if(!moveCountingRegExp.test(fields[5])) {
 		throw new exception.InvalidFEN(fen, i18n.INVALID_MOVE_NUMBER_FIELD);
 	}
-	return { position: position, fiftyMoveClock: parseInt(fields[4], 10), fullMoveNumber: parseInt(fields[5], 10) };
+	return { position: position, fiftyMoveClock: parseInt(fields[4], 10), fullMoveNumber: parseInt(fields[5], 10) || 1 };
```
[the remark this answers is not in the document]
Comment (mbutsykin) [reply src/private_position/fen.js:291]: `pgnRead` completely fails with `Unexpected token` if there are moves with number `0.`

File: test/resources/pgns/misc.pgn [hunk 2]
```diff
@@ -119,11 +119,18 @@ The  value ]}
 [Event "Game with linebreaks in comments"]
 
 1.e4 {[%clk
-01:56:25]} 1...e5 {I am comment
-with a linebreak.}
+01:56:25]} 1...e5 {I am comment\
```
[the remark this answers is not in the document]
Comment (mbutsykin) [reply test/resources/pgns/misc.pgn:122]: Should i move test for this to another place?

[the remark this answers is not in the document]
Comment (yo35) [reply test/resources/pgns/misc.pgn:122]: No, I think the test is fine here. But maybe you can change "I am comment\ with a linebreak and a slash" into something like "I am comment\ with a degenerated backslash followed by a linebreak" in order to highlight more specifically what we want to test here.

File: src/private_position/fen.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (yo35) [reply src/private_position/fen.js:291]: Are you sure? I tried your test-case below, and on my side it passes without this modification.

File: test/resources/pgns/misc.pgn (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbutsykin) [reply test/resources/pgns/misc.pgn:122]: sure no problem)

File: src/private_position/fen.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbutsykin) [reply src/private_position/fen.js:291]: let me check

[the remark this answers is not in the document]
Comment (mbutsykin) [reply src/private_position/fen.js:291]: wow. that is super weird stuff, you are right - it works fine in test w/o my changes.
but i have no idea why lol)

try it for example here
https://npm.runkit.com/kokopu

```
var kokopu = require("kokopu")

const db = kokopu.pgnRead(`
[Event "Game with '0' move number"]
[FEN "8/7R/R3B3/4r1pP/p4k2/8/6PK/r7 w - - 0 0"]

0. Rha7 $2 0-1
`)
```
`db` parsed just fine, but

```
const game  = db.game(0)
```
throws an error `Unrecognized character or group of characters.`
the same error will be throw if you try to parse game (the same in my app)
```
var kokopu = require("kokopu")

const db = kokopu.pgnRead(`
[Event "Game with '0' move number"]
[FEN "8/7R/R3B3/4r1pP/p4k2/8/6PK/r7 w - - 0 0"]

0. Rha7 $2 0-1
`, 0)
```
but in tests - it's fine

[the remark this answers is not in the document]
Comment (mbutsykin) [reply src/private_position/fen.js:291]: can you please take a look and let me know what do you think?

File: test/resources/pgns/misc.pgn (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbutsykin) [reply test/resources/pgns/misc.pgn:122]: updated)

File: src/private_position/fen.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (yo35) [reply src/private_position/fen.js:291]: With your fix on the regular expression `_matchMoveNumber` in tokenstream.js (and without any change in fen.js), both
```
const db = kokopu.pgnRead(`
[Event "Game with '0' move number"]
[FEN "8/7R/R3B3/4r1pP/p4k2/8/6PK/r7 w - - 0 0"]

0. Rha7 $2 0-1
`);
db.game(0);
```
and
```
kokopu.pgnRead(`
[Event "Game with '0' move number"]
[FEN "8/7R/R3B3/4r1pP/p4k2/8/6PK/r7 w - - 0 0"]

0. Rha7 $2 0-1
`, 0)
```
works fine on my side. I guess that in https://npm.runkit.com/kokopu, the engine loads the version of the lib currently in production, which indeed fails with move number 0. It has to be fixed, but the fix is just the modification of `_matchMoveNumber` in tokenstream.js, not what is done in fen.js.

File: test/resources/pgns/misc.pgn (hunk 2 above)
[the remark this answers is not in the document]
Comment (yo35) [reply test/resources/pgns/misc.pgn:122]: thanks

**Comment by yo35** [comment]
Thanks for the fixes. I'll add your name to the list of contributors, and I will issue a new version of the lib this afternoon.

**Comment by mbutsykin** [comment]
Thanks for fast feedback!

**Comment by yo35** [comment]
By the way, I suspected that there could be a [ReDoS](https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS) vulnerability due to your change on the regex `_commentMode`, and it appears to be the case. If you try to parse the following PGN:
```
[Event "ReDoS using comment"]

1. e4 {Evil unterminated comment with lots of backslashes \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\
```
... you'll see that it takes several seconds (and in fact the runtime doubles with each new backslash). I've fixed that in https://github.com/yo35/kokopu/commit/ce236369a636bb10a731ff7a0d4dfad2a11b8409

Regular expressions are sneaky guys! ;-)

**Comment by mbutsykin** [comment]
Wow. nice catch) good job!

**Pull request merged**